File: crates/deepsec-processor/src/process.rs
use crate::agents::{
    AgentBackend, InvestigateBatch, InvestigateFile, ProducedFinding,
};
use crate::batch::batch_records;
use crate::errors::ProcessorError;
use deepsec_core::ids::{generate_run_id, now_iso};
use deepsec_core::store::{load_all_file_records, read_file_record, write_file_record};
use deepsec_core::{
    AnalysisEntry, AnalysisPhase, DataRoot, FileRecord, FileStatus, Finding, InvocationMode,
    ProcessorConfig, RunPhase, RunType, Usage as CoreUsage, complete_run, create_run_meta,
    write_run_meta,
};
use deepsec_scanner::DetectedTech;
use indexmap::IndexMap;
use std::collections::HashSet;
use std::path::PathBuf;

pub struct ProcessOptions {
    pub project_id: String,
    pub project_root: PathBuf,
    pub data_root: DataRoot,
    pub backend: Box<dyn AgentBackend>,
    pub batch_size: usize,
    pub limit: Option<usize>,
    pub filter_prefix: Option<String>,
    pub only_slugs: Vec<String>,
    pub skip_slugs: Vec<String>,
    pub project_info: Option<String>,
    pub prompt_append: Option<String>,
    pub detected: Option<DetectedTech>,
}

#[derive(Debug, Clone, Default)]
pub struct ProcessOutcome {
    pub run_id: String,
    pub batches_run: usize,
    pub analysis_count: usize,
    pub finding_count: usize,
    pub error_batch_count: usize,
    pub quota_exhausted: bool,
}

pub async fn run_process(opts: ProcessOptions) -> Result<ProcessOutcome, ProcessorError> {
    let run_id = generate_run_id();
    let mut meta = create_run_meta(
        &opts.project_id,
        &run_id,
        opts.project_root.to_string_lossy().as_ref(),
        RunType::Process,
    );
    meta.processor_config = Some(ProcessorConfig {
        agent_type: opts.backend.kind().as_str().into(),
        model: opts.backend.model().into(),
        model_config: IndexMap::new(),
        invocation_mode: Some(InvocationMode::Scan),
        source: None,
    });
    write_run_meta(&opts.data_root, &meta)?;

    let records = load_all_file_records(&opts.data_root, &opts.project_id)
        .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;

    let only: HashSet<&str> = opts.only_slugs.iter().map(String::as_str).collect();
    let skip: HashSet<&str> = opts.skip_slugs.iter().map(String::as_str).collect();

    let mut work: Vec<FileRecord> = records
        .into_iter()
        .filter(|r| !r.candidates.is_empty())
        .filter(|r| matches!(r.status, FileStatus::Pending | FileStatus::Error))
        .filter(|r| {
            opts.filter_prefix
                .as_deref()
                .map(|p| r.file_path.starts_with(p))
                .unwrap_or(true)
        })
        .filter(|r| {
            if !only.is_empty() {
                r.candidates.iter().any(|c| only.contains(c.vuln_slug.as_str()))
            } else {
                true
            }
        })
        .filter(|r| {
            if skip.is_empty() {
                return true;
            }
            r.candidates.iter().any(|c| !skip.contains(c.vuln_slug.as_str()))
        })
        .collect();

    // priority: precise candidates first, then by count
    work.sort_by_key(|r| (r.candidates.len() == 0, r.file_path.clone()));

    if let Some(lim) = opts.limit {
        work.truncate(lim);
    }

    let tech_tags = opts
        .detected
        .as_ref()
        .map(|d| d.tags.clone())
        .unwrap_or_default();

    let mut outcome = ProcessOutcome {
        run_id: run_id.clone(),
        ..Default::default()
    };

    let batches = batch_records(&work, opts.batch_size.max(1));
    outcome.batches_run = batches.len();

    let mut total_findings = 0usize;
    let mut total_input = 0u64;
    let mut total_output = 0u64;
    let mut total_cost = 0.0f64;
    let mut total_duration = 0u64;

    for batch in batches {
        let invoke_batch = build_invoke_batch(
            &opts.project_root,
            &batch,
            &tech_tags,
            opts.project_info.as_deref(),
            opts.prompt_append.as_deref(),
        );

        // mark locked
        for r in &batch {
            if let Some(mut rec) = read_file_record(&opts.data_root, &opts.project_id, &r.file_path)
                .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?
            {
                rec.status = FileStatus::Processing;
                rec.locked_by_run_id = Some(run_id.clone());
                rec.locked_at = Some(now_iso());
                write_file_record(&opts.data_root, &rec)
                    .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;
            }
        }

        let result = opts.backend.investigate(&invoke_batch).await;
        match result {
            Ok(out) => {
                total_input += out.usage.input_tokens;
                total_output += out.usage.output_tokens;
                total_cost += out.cost_usd;
                total_duration += out.duration_ms;

                let per_file_cost = if !batch.is_empty() {
                    out.cost_usd / batch.len() as f64
                } else {
                    0.0
                };
                let per_file_usage = CoreUsage {
                    input_tokens: out.usage.input_tokens / batch.len().max(1) as u64,
                    output_tokens: out.usage.output_tokens / batch.len().max(1) as u64,
                    cache_read_input_tokens: out.usage.cache_read_input_tokens
                        / batch.len().max(1) as u64,
                    cache_creation_input_tokens: out.usage.cache_creation_input_tokens
                        / batch.len().max(1) as u64,
                };
                let per_file_duration = out.duration_ms / batch.len().max(1) as u64;

                let results_by_file: std::collections::HashMap<String, Vec<ProducedFinding>> =
                    out.results
                        .into_iter()
                        .map(|r| (r.file_path, r.findings))
                        .collect();

                for r in &batch {
                    let Some(mut rec) =
                        read_file_record(&opts.data_root, &opts.project_id, &r.file_path)
                            .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?
                    else {
                        continue;
                    };
                    let findings = results_by_file.get(&r.file_path).cloned().unwrap_or_default();

                    apply_findings(&mut rec, &run_id, findings.clone());

                    let entry = AnalysisEntry {
                        run_id: run_id.clone(),
                        investigated_at: now_iso(),
                        duration_ms: per_file_duration,
                        duration_api_ms: None,
                        agent_type: opts.backend.kind().as_str().into(),
                        model: opts.backend.model().into(),
                        model_config: IndexMap::new(),
                        agent_session_id: None,
                        finding_count: findings.len(),
                        num_turns: Some(out.num_turns),
                        phase: Some(AnalysisPhase::Process),
                        cost_usd: Some(per_file_cost),
                        usage: Some(per_file_usage.clone()),
                        refusal: None,
                        codex_stderr: None,
                        reinvestigate_marker: None,
                    };
                    rec.analysis_history.push(entry);
                    rec.status = FileStatus::Analyzed;
                    rec.locked_by_run_id = None;
                    rec.locked_at = None;
                    total_findings += findings.len();
                    write_file_record(&opts.data_root, &rec)
                        .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;
                    outcome.analysis_count += 1;
                }
            }
            Err(ProcessorError::Quota(q)) => {
                tracing::warn!("quota exhausted on {}: {}", q.backend, q.detail);
                outcome.quota_exhausted = true;
                outcome.error_batch_count += 1;
                for r in &batch {
                    if let Some(mut rec) =
                        read_file_record(&opts.data_root, &opts.project_id, &r.file_path)
                            .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?
                    {
                        rec.status = FileStatus::Pending;
                        rec.locked_by_run_id = None;
                        rec.locked_at = None;
                        write_file_record(&opts.data_root, &rec)
                            .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;
                    }
                }
                break;
            }
            Err(e) => {
                tracing::error!("batch failed: {e}");
                outcome.error_batch_count += 1;
                for r in &batch {
                    if let Some(mut rec) =
                        read_file_record(&opts.data_root, &opts.project_id, &r.file_path)
                            .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?
                    {
                        rec.status = FileStatus::Error;
                        rec.locked_by_run_id = None;
                        rec.locked_at = None;
                        write_file_record(&opts.data_root, &rec)
                            .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;
                    }
                }
            }
        }
    }

    outcome.finding_count = total_findings;
    meta.stats.files_processed = Some(outcome.analysis_count);
    meta.stats.findings_count = Some(total_findings);
    meta.stats.total_input_tokens = Some(total_input);
    meta.stats.total_output_tokens = Some(total_output);
    meta.stats.total_cost_usd = Some(total_cost);
    meta.stats.total_duration_ms = Some(total_duration);
    write_run_meta(&opts.data_root, &meta)?;
    let final_phase = if outcome.error_batch_count > 0 && outcome.analysis_count == 0 {
        RunPhase::Error
    } else {
        RunPhase::Done
    };
    complete_run(&opts.data_root, &opts.project_id, &run_id, final_phase)
        .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;
    Ok(outcome)
}

fn build_invoke_batch(
    root: &PathBuf,
    batch: &[&FileRecord],
    tech_tags: &[String],
    project_info: Option<&str>,
    prompt_append: Option<&str>,
) -> InvestigateBatch {
    let mut files = Vec::with_capacity(batch.len());
    let mut slug_set: indexmap::IndexSet<String> = indexmap::IndexSet::new();
    for r in batch {
        let content = fs_err::read_to_string(root.join(&r.file_path)).unwrap_or_default();
        for c in &r.candidates {
            slug_set.insert(c.vuln_slug.clone());
        }
        files.push(InvestigateFile {
            path: r.file_path.clone(),
            content,
            candidates: r.candidates.clone(),
        });
    }
    InvestigateBatch {
        project_root: root.clone(),
        files,
        project_info: project_info.map(str::to_string),
        prompt_append: prompt_append.map(str::to_string),
        tech_tags: tech_tags.to_vec(),
        slug_notes: slug_set.into_iter().map(|s| (s, String::new())).collect(),
    }
}

fn apply_findings(rec: &mut FileRecord, run_id: &str, findings: Vec<ProducedFinding>) {
    let mut existing: HashSet<String> = rec
        .findings
        .iter()
        .map(|f| format!("{}|{}", f.vuln_slug, f.title))
        .collect();
    for f in findings {
        let key = format!("{}|{}", f.vuln_slug, f.title);
        if !existing.insert(key) {
            continue;
        }
        rec.findings.push(Finding {
            severity: f.severity,
            vuln_slug: f.vuln_slug,
            title: f.title,
            description: f.description,
            line_numbers: f.line_numbers,
            recommendation: f.recommendation,
            confidence: f.confidence,
            triage: None,
            revalidation: None,
            produced_by_run_id: Some(run_id.to_string()),
        });
    }
}

File: internal/processor/process.go
package processor

import (
	"context"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"sync"

	"github.com/noeljackson/deepsec/internal/core"
	"github.com/noeljackson/deepsec/internal/scanner"
	"golang.org/x/sync/errgroup"
	"golang.org/x/sync/semaphore"
)

// ProcessOptions bundles everything `Process` needs.
type ProcessOptions struct {
	ProjectID         string
	ProjectRoot       string
	DataRoot          core.DataRoot
	Backend           AgentBackend
	ProviderName      string
	BatchSize         int
	Concurrency       int
	Limit             int // 0 = no limit
	FilterPrefix      string
	OnlySlugs         []string
	SkipSlugs         []string
	ProjectInfo       string
	PromptAppend      string
	DirectFiles       []string // direct mode: bypass status filter
	DirectSource      string
	ReinvestigateMark int // 0 = no marker
	MaxCostUSD        float64
	ToolsEnabled      bool
	MaxTurns          int
	SkepticEnabled    bool
	Detected          *scanner.DetectedTech
	// ModelSettings captures the pinned inference knobs (temperature,
	// top-p, seed) for this run. Persisted in
	// ProcessorConfig.ModelConfig and replayed on each AnalysisEntry so
	// processor evaluations are reproducible.
	ModelSettings ModelSettings
}

// ProcessOutcome summarizes one Process run.
type ProcessOutcome struct {
	RunID             string
	BatchesRun        int
	AnalysisCount     int
	FindingCount      int
	ErrorBatchCount   int
	QuotaExhausted    bool
	BudgetExhausted   bool
	TotalCostUSD      float64
	TotalInputTokens  uint64
	TotalOutputTokens uint64
	TotalDurationMs   uint64
}

// Process runs the AI-investigation pipeline over a project's
// FileRecords. Concurrency is bounded by opts.Concurrency; quota errors
// flip a shared cancel flag so in-flight batches stop cleanly.
func Process(ctx context.Context, opts ProcessOptions) (*ProcessOutcome, error) {
	runID := core.GenerateRunID()
	directMode := len(opts.DirectFiles) > 0
	invocationMode := core.InvocationModeScan
	if directMode {
		invocationMode = core.InvocationModeDirect
	}
	meta := core.NewRunMeta(opts.ProjectID, runID, opts.ProjectRoot, core.RunTypeProcess)
	meta.ProcessorConfig = &core.ProcessorConfig{
		AgentType:      opts.ProviderName,
		Model:          opts.Backend.Model(),
		ModelConfig:    opts.ModelSettings.AsMap(),
		InvocationMode: invocationMode,
		Source:         opts.DirectSource,
	}
	if err := opts.DataRoot.WriteRunMeta(meta); err != nil {
		return nil, err
	}

	records, err := opts.DataRoot.LoadAllFileRecords(opts.ProjectID)
	if err != nil {
		return nil, err
	}

	work := filterWork(records, opts, directMode)
	if opts.Limit > 0 && len(work) > opts.Limit {
		work = work[:opts.Limit]
	}

	techTags := []string{}
	if opts.Detected != nil {
		techTags = opts.Detected.Tags
	}

	batches := scanner.BatchRecords(work, batchSize(opts.BatchSize))
	outcome := &ProcessOutcome{
		RunID:      runID,
		BatchesRun: len(batches),
	}

	// Lock everything up front so a parallel run won't claim the same
	// files. Lock failure is non-fatal — log and continue.
	now := core.NowISO()
	for _, batch := range batches {
		for _, r := range batch {
			rec, err := opts.DataRoot.ReadFileRecord(opts.ProjectID, r.FilePath)
			if err != nil || rec == nil {
				continue
			}
			rec.Status = core.StatusProcessing
			rec.LockedByRunID = runID
			rec.LockedAt = now
			_ = opts.DataRoot.WriteFileRecord(rec)
		}
	}

	sem := semaphore.NewWeighted(int64(maxConcurrency(opts.Concurrency)))
	cancelCtx, cancel := context.WithCancel(ctx)
	defer cancel()
	g, ggCtx := errgroup.WithContext(cancelCtx)

	var mu sync.Mutex
	type batchResult struct {
		paths   []string
		out     *InvestigateOutput
		quotaed bool
		failed  bool
		errMsg  string
	}
	results := make([]batchResult, len(batches))

	for i, batch := range batches {
		i, batch := i, batch
		paths := make([]string, len(batch))
		for j, r := range batch {
			paths[j] = r.FilePath
		}
		ibatch := buildInvestigateBatch(opts.ProjectRoot, batch, techTags, opts.ProjectInfo, opts.PromptAppend)
		ibatch.ToolsEnabled = opts.ToolsEnabled
		ibatch.MaxTurns = opts.MaxTurns
		ibatch.MaxCostUSD = opts.MaxCostUSD
		g.Go(func() error {
			if err := sem.Acquire(ggCtx, 1); err != nil {
				// Context cancelled before we got a permit — treat as
				// cancelled, not as an error. Locks go back to pending.
				results[i] = batchResult{paths: paths, quotaed: true}
				return nil
			}
			defer sem.Release(1)
			out, err := opts.Backend.Investigate(ggCtx, ibatch)
			if err != nil {
				if IsQuotaErr(err) {
					mu.Lock()
					outcome.QuotaExhausted = true
					mu.Unlock()
					cancel()
					results[i] = batchResult{paths: paths, quotaed: true}
					return nil
				}
				// A cancelled context (sibling batch quotaed-out, or the
				// caller cancelled) shouldn't downgrade these files to
				// status=error; release them back to pending instead.
				if errors.Is(err, context.Canceled) {
					results[i] = batchResult{paths: paths, quotaed: true}
					return nil
				}
				results[i] = batchResult{paths: paths, failed: true, errMsg: err.Error()}
				return nil
			}
			mu.Lock()
			outcome.TotalCostUSD += out.CostUSD
			outcome.TotalInputTokens += out.Usage.InputTokens
			outcome.TotalOutputTokens += out.Usage.OutputTokens
			outcome.TotalDurationMs += out.DurationMs
			if opts.MaxCostUSD > 0 && outcome.TotalCostUSD >= opts.MaxCostUSD {
				outcome.BudgetExhausted = true
				cancel()
			}
			mu.Unlock()
			results[i] = batchResult{paths: paths, out: out}
			return nil
		})
	}
	_ = g.Wait()

	// Apply results sequentially so concurrent writers can't fight over
	// the same FileRecord. The order is deterministic and parsing-cheap.
	for _, r := range results {
		switch {
		case r.quotaed:
			outcome.ErrorBatchCount++
			releaseLocks(opts.DataRoot, opts.ProjectID, r.paths, core.StatusPending)
		case r.failed:
			outcome.ErrorBatchCount++
			releaseLocks(opts.DataRoot, opts.ProjectID, r.paths, core.StatusError)
		case r.out == nil:
			// cancelled before dispatch; release back to pending
			releaseLocks(opts.DataRoot, opts.ProjectID, r.paths, core.StatusPending)
		default:
			if opts.SkepticEnabled {
				if cost, err := applySkeptic(ctx, opts, r.out); err == nil {
					outcome.TotalCostUSD += cost
					if opts.MaxCostUSD > 0 && outcome.TotalCostUSD >= opts.MaxCostUSD {
						outcome.BudgetExhausted = true
					}
				}
			}
			applied := applyBatch(opts, runID, r.paths, r.out)
			outcome.AnalysisCount += applied.analyzed
			outcome.FindingCount += applied.findings
		}
	}

	fp := outcome.AnalysisCount
	meta.Stats.FilesProcessed = &fp
	meta.Stats.FindingsCount = &outcome.FindingCount
	meta.Stats.TotalInputTokens = &outcome.TotalInputTokens
	meta.Stats.TotalOutputTokens = &outcome.TotalOutputTokens
	meta.Stats.TotalCostUSD = &outcome.TotalCostUSD
	meta.Stats.TotalDurationMs = &outcome.TotalDurationMs
	if err := opts.DataRoot.WriteRunMeta(meta); err != nil {
		return nil, err
	}
	phase := core.RunPhaseDone
	if outcome.ErrorBatchCount > 0 && outcome.AnalysisCount == 0 {
		phase = core.RunPhaseError
	}
	if _, err := opts.DataRoot.CompleteRun(opts.ProjectID, runID, phase); err != nil {
		return nil, err
	}
	return outcome, nil
}

type applyStats struct {
	analyzed int
	findings int
}

func applyBatch(opts ProcessOptions, runID string, paths []string, out *InvestigateOutput) applyStats {
	stats := applyStats{}
	n := len(paths)
	if n == 0 {
		return stats
	}
	perCost := out.CostUSD / float64(n)
	perUsage := core.Usage{
		InputTokens:              out.Usage.InputTokens / uint64(n),
		OutputTokens:             out.Usage.OutputTokens / uint64(n),
		CacheReadInputTokens:     out.Usage.CacheReadInputTokens / uint64(n),
		CacheCreationInputTokens: out.Usage.CacheCreationInputTokens / uint64(n),
	}
	perDur := out.DurationMs / uint64(n)

	byFile := map[string][]ProducedFinding{}
	for _, r := range out.Results {
		byFile[r.FilePath] = r.Findings
	}

	for _, path := range paths {
		rec, err := opts.DataRoot.ReadFileRecord(opts.ProjectID, path)
		if err != nil || rec == nil {
			continue
		}
		findings := byFile[path]
		if out.Refusal != nil {
			rec.AnalysisHistory = append(rec.AnalysisHistory, core.AnalysisEntry{
				RunID:          runID,
				InvestigatedAt: core.NowISO(),
				DurationMs:     perDur,
				AgentType:      opts.ProviderName,
				Model:          opts.Backend.Model(),
				ModelConfig:    opts.ModelSettings.AsMap(),
				FindingCount:   0,
				Phase:          core.PhaseProcess,
				CostUSD:        ptr(perCost),
				Usage:          &perUsage,
				Refusal:        out.Refusal,
			})
			rec.Status = core.StatusError
			rec.LockedByRunID = ""
			rec.LockedAt = ""
			_ = opts.DataRoot.WriteFileRecord(rec)
			continue
		}
		appendFindings(rec, runID, findings)
		entry := core.AnalysisEntry{
			RunID:          runID,
			InvestigatedAt: core.NowISO(),
			DurationMs:     perDur,
			AgentType:      opts.ProviderName,
			Model:          opts.Backend.Model(),
			ModelConfig:    opts.ModelSettings.AsMap(),
			FindingCount:   len(findings),
			NumTurns:       intPtr(out.NumTurns),
			Phase:          core.PhaseProcess,
			CostUSD:        ptr(perCost),
			Usage:          &perUsage,
		}
		if opts.ReinvestigateMark > 0 {
			entry.ReinvestigateMark = intPtr(opts.ReinvestigateMark)
		}
		rec.AnalysisHistory = append(rec.AnalysisHistory, entry)
		rec.Status = core.StatusAnalyzed
		rec.LockedByRunID = ""
		rec.LockedAt = ""
		stats.findings += len(findings)
		stats.analyzed++
		_ = opts.DataRoot.WriteFileRecord(rec)
	}
	return stats
}

func appendFindings(rec *core.FileRecord, runID string, findings []ProducedFinding) {
	seen := map[string]bool{}
	for _, f := range rec.Findings {
		seen[f.VulnSlug+"|"+f.Title] = true
	}
	for _, f := range findings {
		key := f.VulnSlug + "|" + f.Title
		if seen[key] {
			continue
		}
		seen[key] = true
		rec.Findings = append(rec.Findings, core.Finding{
			Severity:        f.Severity,
			VulnSlug:        f.VulnSlug,
			Title:           f.Title,
			Description:     f.Description,
			LineNumbers:     f.LineNumbers,
			Recommendation:  f.Recommendation,
			Confidence:      f.Confidence,
			ProducedByRunID: runID,
		})
	}
}

func releaseLocks(root core.DataRoot, projectID string, paths []string, status core.FileStatus) {
	for _, p := range paths {
		rec, err := root.ReadFileRecord(projectID, p)
		if err != nil || rec == nil {
			continue
		}
		rec.Status = status
		rec.LockedByRunID = ""
		rec.LockedAt = ""
		_ = root.WriteFileRecord(rec)
	}
}

func filterWork(records []*core.FileRecord, opts ProcessOptions, directMode bool) []*core.FileRecord {
	onlySet := toStringSet(opts.OnlySlugs)
	skipSet := toStringSet(opts.SkipSlugs)
	directSet := toStringSet(opts.DirectFiles)

	out := make([]*core.FileRecord, 0)
	for _, r := range records {
		if directMode {
			if _, ok := directSet[r.FilePath]; !ok {
				continue
			}
		} else {
			if len(r.Candidates) == 0 {
				continue
			}
			if r.Status != core.StatusPending && r.Status != core.StatusError {
				continue
			}
		}
		if opts.FilterPrefix != "" && !hasPrefix(r.FilePath, opts.FilterPrefix) {
			continue
		}
		if len(onlySet) > 0 && !hasAny(r.Candidates, onlySet) {
			continue
		}
		if len(skipSet) > 0 && allIn(r.Candidates, skipSet) {
			continue
		}
		if opts.ReinvestigateMark > 0 && hasWave(r, opts.ReinvestigateMark, opts.ProviderName) {
			continue
		}
		out = append(out, r)
	}
	sort.SliceStable(out, func(i, j int) bool {
		ai, aj := len(out[i].Candidates) == 0, len(out[j].Candidates) == 0
		if ai != aj {
			return !ai
		}
		return out[i].FilePath < out[j].FilePath
	})
	return out
}

func hasWave(r *core.FileRecord, wave int, agent string) bool {
	for _, e := range r.AnalysisHistory {
		if e.ReinvestigateMark != nil && *e.ReinvestigateMark == wave &&
			e.AgentType == agent && e.Phase != core.PhaseRevalidate {
			return true
		}
	}
	return false
}

func hasAny(cs []core.CandidateMatch, set map[string]struct{}) bool {
	for _, c := range cs {
		if _, ok := set[c.VulnSlug]; ok {
			return true
		}
	}
	return false
}

func allIn(cs []core.CandidateMatch, set map[string]struct{}) bool {
	for _, c := range cs {
		if _, ok := set[c.VulnSlug]; !ok {
			return false
		}
	}
	return true
}

func toStringSet(xs []string) map[string]struct{} {
	out := map[string]struct{}{}
	for _, x := range xs {
		if x != "" {
			out[x] = struct{}{}
		}
	}
	return out
}

func hasPrefix(s, p string) bool {
	return len(s) >= len(p) && s[:len(p)] == p
}

func buildInvestigateBatch(root string, batch []*core.FileRecord, techTags []string, info, append_ string) *InvestigateBatch {
	files := make([]InvestigateFile, 0, len(batch))
	slugSet := map[string]struct{}{}
	for _, r := range batch {
		body, _ := os.ReadFile(filepath.Join(root, r.FilePath))
		for _, c := range r.Candidates {
			slugSet[c.VulnSlug] = struct{}{}
		}
		files = append(files, InvestigateFile{
			Path:       r.FilePath,
			Content:    string(body),
			Candidates: r.Candidates,
		})
	}
	slugs := make([]string, 0, len(slugSet))
	for k := range slugSet {
		slugs = append(slugs, k)
	}
	return &InvestigateBatch{
		ProjectRoot:  root,
		Files:        files,
		ProjectInfo:  info,
		PromptAppend: append_,
		TechTags:     techTags,
		SlugNotes:    slugs,
	}
}

func batchSize(n int) int {
	if n < 1 {
		return 5
	}
	return n
}

func maxConcurrency(n int) int {
	if n < 1 {
		return 4
	}
	return n
}

func ptr(f float64) *float64 { return &f }
func intPtr(i int) *int      { return &i }

// LoadAllFileRecords is exposed here so the CLI can reuse it without
// importing core directly for one read path.
func LoadAllFileRecords(root core.DataRoot, projectID string) ([]*core.FileRecord, error) {
	out, err := root.LoadAllFileRecords(projectID)
	if err != nil && !errors.Is(err, fs.ErrNotExist) {
		return nil, fmt.Errorf("loading file records: %w", err)
	}
	return out, nil
}
